frontend/src/api/prometheus/serving.ts

```diff
@@ -0,0 +1,94 @@
+import * as React from 'react';
+import { ContextResourceData, PrometheusQueryRangeResultValue } from '../../types';
+import { useContextResourceData } from '../../utilities/useContextResourceData';
+import { ModelServingMetricType } from '../../pages/modelServing/screens/metrics/ModelServingMetricsContext';
+import usePrometheusQueryRange from './usePrometheusQueryRange';
+import { TimeframeTitle } from '../../pages/modelServing/screens/types';
+import { TimeframeStep, TimeframeTime } from '../../pages/modelServing/screens/const';
+
+export const useModelServingMetrics = (
+  queries: Record<ModelServingMetricType, string>,
+  timeframe: TimeframeTitle,
+  lastUpdateTime: number,
+  setLastUpdateTime: (time: number) => void,
+): {
+  data: Record<ModelServingMetricType, ContextResourceData<PrometheusQueryRangeResultValue>>;
+  refresh: () => void;
+} => {
+  const now = Date.now();
+  let end: number;
+  // every 5 minutes refetch automatically
+  if (now - lastUpdateTime > 5 * 60 * 1000) {
+    end = now;
+    setLastUpdateTime(now);
```

Comment:
woah -- no mutations in the render thread! Do this in an effect if you have to do it.